etl/steps/data/garden/war/2023-08-05/cow_mid.py

```diff
@@ -125,39 +99,103 @@ def run(dest_dir: str) -> None:
+    """Process MIDB table.
+
+    - Sanity checks
+    - Keep relevant columns
+    - Add observation per year
+    - Add regions
     """
-    # Reduce NaNs in `fatalpre`
-    fatality_lower_bound = {
-        0: 0,
-        1: 1,
-        2: 26,
-        3: 101,
-        4: 251,
-        5: 501,
-        6: 1000,
-        -9: np.nan,
-    }
-    mask = tb["fatalpre"] == -9
-    tb.loc[mask, "fatalpre"] = tb.loc[mask, "fatality"].map(fatality_lower_bound)
-
-    assert tb["fatalpre"].isna().sum() == 599, "Different number of NaNs found. Expected was 599."
-    tb["fatalpre"] = tb["fatalpre"].fillna(0)
+    # Sanity checks
+    assert (
+        tb.groupby(["dispnum", "ccode", "styear", "endyear"]).size().max() == 1
+    ), "Multiple entries for a conflict-country-start_year-end_year"
+    assert tb["styear"].notna().all() and (tb["styear"] >= 0).all(), "NA values (or negative) found in `styear`"
+    assert tb["endyear"].notna().all() and (tb["endyear"] >= 0).all(), "NA values (or negative) found in `endyear`"
+
+    # Add regions
+    tb = add_regions(tb)
+
+    # Keep relevant columns
+    COLUMNS_RELEVANT = [
+        "dispnum",
+        "styear",
+        "endyear",
+        "region",
+    ]
+    tb = tb[COLUMNS_RELEVANT]
+
+    # Drop duplicates
+    tb = tb.drop_duplicates()
+
+    # Add observation for each year
+    tb = expand_observations(tb)
+
+    # Drop columns
+    tb = tb.drop(columns=["styear", "endyear"])
 
     return tb
 
 
-def _sanity_checks(tb: Table) -> Table:
-    # sanity checks
-    assert tb.groupby(["dispnum", "ccode", "styear", "endyear"]).size().max() == 1
-    assert set(tb["hostlev"]) == {1, 2, 3, 4, 5}, "Set of hostlev values should be {1, 2, 3, 4, 5}."
-    assert ((tb["ccode"] > 1) & (tb["ccode"] < 1000)).all(), "ccode should be between 1 and 1000."
+def combine_tables(tb_a: Table, tb_b: Table) -> Table:
+    """Combine MIDA and MIDB processed tables.
+
+    Basically, it adds region information (from MIDB) to MIDA.
+    """
+    # Merge
+    tb = tb_a.merge(tb_b, on=["dispnum", "year"], how="left")
+
+    # Fill NaNs
+    ## Sanity check (1)
+    dispnum_nans_expected = {2044, 2328, 4005}
+    dispnum_nans_found = set(tb.loc[tb["region"].isna()])
+    dispnum_nans_unexpected = dispnum_nans_found - dispnum_nans_expected
+    assert dispnum_nans_unexpected, f"Unexpected dispnum with NaN regions: {dispnum_nans_unexpected}"
+    ## Sanity check (2)
+    assert (
+        tb_b[tb_b["dispnum"].isin(dispnum_nans_expected)].groupby("dispnum")["region"].nunique().max() == 1
+    ), f"More than one region for some dispnum in {dispnum_nans_expected}"
+    ## Actually fill NaNs
+    tb.loc[tb["dispnum"] == 2044, "region"] = "Americas"
+    tb.loc[tb["dispnum"] == 2328, "region"] = "Europe"
+    tb.loc[tb["dispnum"] == 4005, "region"] = "Asia"
+
+    # Check there is no NaN!
+    assert (tb.isna().sum() == 0).all(), "NaN in some field!"
+
+    return tb
 
 
 def add_regions(tb: Table) -> Table:
```

Comment:
The codes assigned are slightly different to what you defined on the yaml:
- Americas includes 1 here, but the yaml doesn't
- Middle East doesn't include 699 here, but the yaml does

Reply:
Aligning these, good catch!